lib/libc/sys/uuidgen.2
.\" $NetBSD: uuidgen.2,v 1.4 2005/01/30 18:13:04 jwise Exp $
.\" $FreeBSD: src/lib/libc/sys/uuidgen.2,v 1.7 2003/06/27 13:41:29 yar Exp $
.\" Copyright (c) 2002 Marcel Moolenaar
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $FreeBSD: src/lib/libc/sys/uuidgen.2,v 1.7 2003/06/27 13:41:29 yar Exp $
.\"
.Dd May 26, 2002
.Dt UUIDGEN 2
.Os
.Sh NAME
.Nm uuidgen
.Nd generate universally unique identifiers
.Sh SYNOPSIS
.In sys/types.h
.In sys/uuid.h
.Ft int
.Fn uuidgen "struct uuid *store" "int count"
.Sh DESCRIPTION
The
.Fn uuidgen
system call generates
.Fa count
time-based universally unique identifiers (UUIDs) and writes them to the
buffer pointed to by
.Fa store .
.Pp
Also known as globally unique identifiers (GUIDs), UUIDs are composed of
128 bits arranged follows:
.Bd -literal
struct uuid {
	uint32_t	time_low;
	uint16_t	time_mid;
	uint16_t	time_hi_and_version;
	uint8_t		clock_seq_hi_and_reserved;
	uint8_t		clock_seq_low;
	uint8_t		node[UUID_NODE_LEN];
};
.Ed
.Bl -tag -width ".Va clock_seq_hi_and_reserved"
.It Va time_low
The least significant 32 bits of a 60-bit timestamp.
.It Va time_mid
The least significant 16 bits of the most significant 28 bits of the 60-bit
timestamp.
.It Va time_hi_and_version
The most significant 12 bits of the 60-bit timestamp multiplexed with a 4-bit
version number.
The version number is stored in the most significant 4 bits of the 16-bit
field.
.It Va clock_seq_hi_and_reserved
The most significant 6 bits of a 14-bit sequence number multiplexed with a
2-bit variant value.
Identifiers generated by the
.Fn uuidgen
system call have a variant value of 2.
The variant value is stored in the most significant bits of the field.
.It Va clock_seq_low
The least significant 8 bits of a 14-bit sequence number.
.It Va node
A random 6-byte IEEE 802 (MAC) multicast address.
.El
.Pp
All fields are stored in native byte order.
.Pp
If
.Fa count
is bigger than 1, the generated identifiers will be numerically
consecutive.
.Sh RETURN VALUES
Upon successful completion, a value of 0 is returned.
Otherwise, a value of \-1 is returned and
.Va errno
is set to indicate the error.
.Sh ERRORS
The
.Fn uuidgen
system call can fail with:
.Bl -tag -width Er
.It Bq Er EFAULT
The buffer pointed to by
.Fa store
could not be written to for at least one identifier.
.It Bq Er EINVAL
The
.Fa count
argument is less than 1 or larger than 2048.
.El
.Sh SEE ALSO
.Xr uuidgen 1 ,
.Xr uuid 3
.Sh STANDARDS
The identifiers provided by the
.Fn uuidgen
system call are generated and represented in conformance with the DCE
1.1 RPC specification.
The system call itself, however, is not part of the specification.


usr.bin/uuidgen/uuidgen.c
/*	$NetBSD: uuidgen.c,v 1.4 2011/09/16 15:39:30 joerg Exp $	*/

/*
 * Copyright (c) 2004 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Jason R. Thorpe.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Copyright (c) 2002 Marcel Moolenaar
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <uuid.h>

__dead static void
usage(void)
{
	extern char *__progname;

	(void)fprintf(stderr, "usage: %s [-1s] [-n count] [-o filename]\n",
	    __progname);
	exit(1);
}

int
main(int argc, char *argv[])
{
	FILE *fp;
	uuid_t *store, *uuid;
	char *p;
	int ch, count, i, iterate, c_struct;

	count = -1;	/* no count yet */
	fp = stdout;	/* default output file */
	iterate = 0;	/* not one at a time */
	c_struct = 0;	/* not as a C structure */

	while ((ch = getopt(argc, argv, "1n:o:s")) != -1) {
		switch (ch) {
		case '1':
			iterate = 1;
			break;
		case 'n':
			if (count > 0)
				usage();
			count = strtol(optarg, &p, 10);
			if (*p != 0 || count < 1)
				usage();
			break;
		case 'o':
			if (fp != stdout)
				errx(1, "multiple output files not allowed");
			fp = fopen(optarg, "w");
			if (fp == NULL)
				err(1, "fopen");
			break;
		case 's':
			c_struct = 1;
			break;
		default:
			usage();
		}
	}
	argv += optind;
	argc -= optind;

	if (argc)
		usage();

	if (count == -1)
		count = 1;

	store = (uuid_t*)malloc(sizeof(uuid_t) * count);
	if (store == NULL)
		err(1, "malloc()");

	if (!iterate) {
		/* Get them all in a single batch */
		if (uuidgen(store, count) != 0)
			err(1, "uuidgen()");
	} else {
		uuid = store;
		for (i = 0; i < count; i++) {
			if (uuidgen(uuid++, 1) != 0)
				err(1, "uuidgen()");
		}
	}

	uuid = store;
	while (count--) {
		uuid_to_string(uuid++, &p, NULL);
		if (c_struct) {
			fprintf(fp, "= { /* %s */\n", p);	/* } */
			/*
			 * Chunk up the string for processing:
			 *
			 *	aaaaaaaa-bbbb-cccc-dddd-0123456789ab
			 *
			 * We output it like so:
			 *
			 * = { \/\* aaaaaaaa-bbbb-cccc-ddee-0123456789ab \*\/
			 *	0xaaaaaaaa,
			 *	0xbbbb,
			 *	0xcccc,
			 *	0xdd,
			 *	0xee,
			 *	{ 0x01, 0x23, 0x45, 0x67, 0x89, 0xab }
			 * };
			 */
			p[8] = '\0';	/* aaaaaaaa */
			p[13] = '\0';	/* bbbb */
			p[18] = '\0';	/* cccc */
			p[23] = '\0';	/* dddd */
			fprintf(fp, "\t0x%s,\n", p);
			fprintf(fp, "\t0x%s,\n", &p[9]);
			fprintf(fp, "\t0x%s,\n", &p[14]);
			fprintf(fp, "\t0x%c%c,\n", p[19], p[20]);
			fprintf(fp, "\t0x%c%c,\n", p[21], p[22]);
			fprintf(fp, "\t{ 0x%c%c, 0x%c%c, 0x%c%c, 0x%c%c, "
					"0x%c%c, 0x%c%c }\n",
				p[24], p[25], p[26], p[27],
				p[28], p[29], p[30], p[31],
				p[32], p[33], p[34], p[35]);
	/* { */		fprintf(fp, "};\n");
		} else
			fprintf(fp, "%s\n", p);
		free(p);
	}

	free(store);
	if (fp != stdout)
		fclose(fp);
	return (0);
}
